src/abstractdataframe/join.jl

```diff
@@ -81,16 +81,628 @@ end
+        # here we know that x_refpool is AbstractVector that allows integer indexing
+        # and its firstindex must be an integer
+        # if firstindex is not 0 or 1 then we fallback to slow path for safety reasons
+        # all refpool we currently know have firstindex 0 or 1
+        # if there is some very strange firstindex we might run into overflow issues
+        # below use function barrier as mapping is not type stable
+        mapping = [get(invrefpool, v, missing) for v in x_refpool]
+        return map_refarray(mapping, DataAPI.refarray(x), Val(Int(firstindex(x_refpool))))
+    else
+        return [get(invrefpool, v, missing) for v in x]
+    end
+end
+
+function compose_inner_table(joiner::DataFrameJoiner,
+                             makeunique::Bool,
+                             left_rename::Union{Function, AbstractString, Symbol},
+                             right_rename::Union{Function, AbstractString, Symbol})
+
+    right_len = length(joiner.dfr_on[!, 1])
+    left_len = length(joiner.dfl_on[!, 1])
+    right_shorter = right_len <= left_len
+
+    left_cols = collect(eachcol(joiner.dfl_on))
+    right_cols = collect(eachcol(joiner.dfr_on))
+
+    # if column of a longer table supports DataAPI.refpool and DataAPI.invrefpool
+    # remap matching left and right columns to use refs
+    if right_shorter
+        for i in eachindex(left_cols, right_cols)
+            rc = right_cols[i]
+            lc = left_cols[i]
+
+            lc_refs = DataAPI.refarray(lc)
+            lc_refpool = DataAPI.refpool(lc)
+            lc_invrefpool = DataAPI.invrefpool(lc)
+            if check_mapping_allowed(rc, lc_refs, lc_refpool, lc_invrefpool)
+                right_cols[i] = map2refs(rc, lc_invrefpool)
+                left_cols[i] = lc_refs
+            end
+        end
+    else
+        for i in eachindex(left_cols, right_cols)
+            rc = right_cols[i]
+            lc = left_cols[i]
+
+            rc_refs = DataAPI.refarray(rc)
+            rc_refpool = DataAPI.refpool(rc)
+            rc_invrefpool = DataAPI.invrefpool(rc)
+            if check_mapping_allowed(lc, rc_refs, rc_refpool, rc_invrefpool)
+                right_cols[i] = rc_refs
+                left_cols[i] = map2refs(lc, rc_invrefpool)
+            end
+        end
+    end
+
+    # this is a workaround for https://github.com/JuliaData/CategoricalArrays.jl/issues/319
+    # this path will be triggered only in rare cases when the refpool code above
+    # fails to convert CategoricalArray into refpool
+    disallow_sorted = false
+
+    for (lc, rc) in zip(left_cols, right_cols)
+        @assert length(lc) == left_len
+        @assert length(rc) == right_len
+        lct = typeof(lc)
+        rct = typeof(rc)
+        rcat = rct.name === :CategoricalArray && nameof(rct.module) === :CategoricalArrays
+        lcat = lct.name === :CategoricalArray && nameof(lct.module) === :CategoricalArrays
```

Comment:
Actually I've just discovered that this pattern doesn't work on Julia 1.7 as `T.name` no longer gives a symbol. And actually `T.module` doesn't seem to work even on Julia 1.0. Something which seems to work is:
```
nameof(T) === :CategoricalArray && nameof(parentmodule(T)) === :CategoricalArrays
```

Reply:
OK - I will then change the same in the groupeddataframe/fastaggregates.jl as we use the same pattern there.

Reply:
Got it: that's because you need `T.name.name` and `T.name.module`. But better use the public functions instead of internal fields anyway.